salt/vault/roles.sls

```diff
@@ -0,0 +1,18 @@
+{% set roles = salt.pillar.get('vault:roles') %}
+{% for role_id, role in roles.items() %}
+{% for environment in role.environments %}
+{% set environment = salt.pillar.get('environments:{}'.format(environment), {}) %}
+  {% for purpose in env_info.purposes %}
```

Comment:
I started writing one answer and then realized that `env_info` is a remnant of one of my earlier approaches. Good catch :)